ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStoreAdaptor.java

```diff
@@ -0,0 +1,349 @@
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-03
+ */
+public class SlotStoreAdaptor implements ObjectStoreAPI {
+
+    private final SlotStore store;
+
+    // used for standalone bootstrap via StoreManager
+    public SlotStoreAdaptor() throws IOException {
+        SlotStoreEnvironmentBean envBean = BeanPopulator.getDefaultInstance(SlotStoreEnvironmentBean.class);
+        store = new SlotStore(envBean);
+    }
+
+    // used for beans wiring type bootstrap when running embedded.
+    public SlotStoreAdaptor(SlotStore slotStore) {
+        this.store = slotStore;
+    }
+
+    /**
+     * Obtain all of the Uids for a specified type.
+     *
+     * @param typeName       The type to scan for.
+     * @param foundInstances The object state in which to store the Uids
+     * @param matchState     The file type to look for (e.g., committed, shadowed). [StateStatus]
+     * @return <code>true</code> if no errors occurred, <code>false</code>
+     * otherwise.
+     */
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances, int matchState) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ", " + matchState + ")");
+        }
+
+        SlotStoreKey key = new SlotStoreKey(null, typeName, matchState);
+
+        SlotStoreKey[] matchingKeys = store.getMatchingKeys(key);
+
+        OutputObjectState buffer = new OutputObjectState();
+
+        try {
+            if (matchingKeys != null) {
+                for (SlotStoreKey matchingKey : matchingKeys) {
+                    if (matchingKey != null) {
+                        UidHelper.packInto(matchingKey.getUid(), buffer);
+                    }
+                }
+            }
+            UidHelper.packInto(Uid.nullUid(), buffer);
+        } catch (IOException e) {
+            throw new ObjectStoreException(e);
+        }
+
+        foundInstances.setBuffer(buffer.buffer());
+
+        return true;
+    }
+
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ")");
+        }
+
+        return allObjUids(typeName, foundInstances, StateStatus.OS_UNKNOWN);
+    }
+
+    /**
+     * Obtain all types of objects stored in the object store.
+     *
+     * @param foundTypes The state in which to store the types.
+     * @return <code>true</code> if no errors occurred, <code>false</code>
+     * otherwise.
+     */
+    @Override
+    public boolean allTypes(InputObjectState foundTypes) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allTypes()");
+        }
+
+        String[] knownTypes = store.getKnownTypes(); // may contain trailing null elements
```

Comment:
The comment is confusing because the array of `knownTypes` is populated in the method `getKnownTypes` by the call `types.add(key.getTypeName());` but the added typeName can never be null (because the `SlotStoreKey` constructor dereferences it when checks for "/").

Reply:
oops. That's a copy-paste from the Hornetq version, which can do odd things because it's backed by a ConcurrentMap. You're right, it doesn't apply here.